plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,272 @@
+        inPackDescr <- Development.IDE.useWithStale ParseCabalFile $ toNormalizedFilePath cabalFilePath
+        let mbCnfOrigContents = case snd . fst <$> contents of
+                    Just (Just txt) -> Just $ encodeUtf8 txt
+                    _               -> Nothing
+        let mbFields = fst <$> inFields
+        let mbPackDescr :: Maybe GenericPackageDescription = fst <$> inPackDescr
+        pure (mbCnfOrigContents, mbFields, mbPackDescr)
+
+  (cnfOrigContents, fields, packDescr) <- do
+    cnfOrigContents <- case mbCnfOrigContents of
+          (Just cnfOrigContents) -> pure cnfOrigContents
+          Nothing                -> readCabalFile cabalFilePath
+    (fields, packDescr) <- case (mbFields, mbPackDescr) of
+          (Just fields, Just packDescr) -> pure (fields, packDescr)
+          (_, _) -> case parseCabalFile cabalFilePath cnfOrigContents of
+                        Left err   -> throwE $ PluginInternalError $ T.pack $ err
+                        Right (f ,gpd) -> pure (f, gpd)
+    pure (cnfOrigContents, fields, packDescr)
+
+  let inputs = do
+        let rcnfComponent = buildTarget
+        let specVer = specVersion $ packageDescription packDescr
+        cmp <- resolveComponent cabalFilePath (fields, packDescr) rcnfComponent
+        deps <- traverse (validateDependency specVer) dependency
+        pure (fields, packDescr, cmp, deps)
+
+  (cnfFields, origPackDescr, cnfComponent, cnfDependencies) <- case inputs of
+    Left err   -> throwE $ PluginInternalError $ T.pack err
+    Right pair -> pure pair
+
+  case executeConfig (validateChanges origPackDescr) (Config {..}) of
+    Nothing -> throwE $ PluginInternalError $ T.pack $ "Cannot extend build-depends in " ++ cabalFilePath
+    Just newContents  -> do
+              let edit = diffText caps (verTxtDocId, T.decodeUtf8 cnfOrigContents) (T.decodeUtf8 newContents) SkipDeletions
+              Logger.logWith recorder Logger.Info $ LogCreatedEdit edit
+              pure edit
+
+-- | Given a path to a haskell file, returns the closest cabal file.
+--   If cabal file wasn't found, gives Nothing.
+findResponsibleCabalFile :: FilePath -> IO (Maybe FilePath)
+findResponsibleCabalFile haskellFilePath = do
+  let dirPath = dropFileName haskellFilePath
+      allDirPaths = reverse $ scanl1 (</>) (splitPath dirPath) -- sorted from most to least specific
+  go allDirPaths
+  where
+    go [] = pure Nothing
+    go (path:ps) = do
+      objects <- listDirectory path
+      let objectsWithPaths = map (\obj -> path <> obj) objects
+          objectsCabalExtension = filter (\c -> takeExtension c == ".cabal") objectsWithPaths
+      cabalFiles <- filterM (\c -> doesFileExist c) objectsCabalExtension
+      case safeHead cabalFiles of
+        Nothing        -> go ps
+        Just cabalFile -> pure $ Just cabalFile
+
+-- | Gives cabal file's contents or throws error.
+--   Inspired by @readCabalFile@ in cabal-add,
+--   Distribution.Client.Main
```

Comment:
Please add this is a *fallback* path and not how you should read the .cabal file in the average case.